src/audio/tdfb/tdfb_direction.c

```diff
@@ -0,0 +1,555 @@
+ * \param[in]		x		Input PCM sample
+ */
+void tdfb_direction_copy_emphasis(struct tdfb_comp_data *cd, int ch_count, int *ch, int32_t x)
+{
+	int32_t y;
+
+	if (!cd->direction_updates)
+		return;
+
+	y = iir_df2t(&cd->direction.emphasis[*ch], x);
+	*cd->direction.wp = sat_int16(Q_SHIFT_RND(y, 31, 17)); /* 12 dB boost after high-pass */
+	cd->direction.wp++;
+	tdfb_cinc_s16(&cd->direction.wp, cd->direction.d_end, cd->direction.d_size);
+	(*ch)++;
+	if (*ch == ch_count)
+		*ch = 0;
+}
+
+/* Use function y = sqrt_int16(x), where x is Q4.12 unsigned, y is Q4.12
+ * TODO: Add scaling of input to Q4.12 with a suitable N^2 or (1/N)^2 value
+ * to increase sqrt() computation precision and avoid MIN() function to
+ * saturate for large distances (over 4m). The current virtual source
+ * distance 3m allows max 1m size array for this to work correctly without
+ * hitting the saturation.
+ */
+static inline int16_t tdfb_mic_distance_sqrt(int32_t x)
+{
+	int32_t xs;
+
+	xs = Q_SHIFT_RND(x, 24, 12);
+	xs = MIN(xs, UINT16_MAX);
+	return sqrt_int16((uint16_t)xs);
+}
+
+static int16_t max_mic_distance(struct tdfb_comp_data *cd)
+{
+	int16_t d;
+	int16_t dx;
+	int16_t dy;
+	int16_t dz;
+	int32_t d2;
+	int32_t d2_max = 0;
+	int i;
+	int j;
+
+	/* Max lag based on largest array dimension. Microphone coordinates are Q4.12 meters */
+	for (i = 0; i < cd->config->num_mic_locations; i++) {
+		for (j = 0; i < cd->config->num_mic_locations; i++) {
+			if (j == i)
+				continue;
+
+			dx = cd->mic_locations[i].x - cd->mic_locations[j].x;
+			dy = cd->mic_locations[i].y - cd->mic_locations[j].y;
+			dz = cd->mic_locations[i].z - cd->mic_locations[j].z;
+
+			/* d2 is Q8.24 meters */
+			d2 = dx * dx + dy * dy + dz * dz;
+			if (d2 > d2_max)
+				d2_max = d2;
+		}
+	}
+
+	/* Squared distance is Q8.24, return Q4.12 meters */
+	d = tdfb_mic_distance_sqrt(d2_max);
+	return d;
+}
+
+static bool line_array_mode_check(struct tdfb_comp_data *cd)
+{
+	int32_t px, py, pz;
+	int16_t a, b, c, d, e, f;
+	int i;
+	int num_mic_locations = cd->config->num_mic_locations;
+
+	if (num_mic_locations == 2)
+		return true;
+
+	/* Cross product of vectors AB and AC is (0, 0, 0) if they are co-linear.
+	 * Form vector AB(a,b,c) from x(i+1) - x(i), y(i+1) - y(i), z(i+1) - z(i)
+	 * Form vector AC(d,e,f) from x(i+2) - x(i), y(i+2) - y(1), z(i+2) - z(i)
+	 * Cross product AB x AC is (b * f - c * e, c * d - a * f, a * e - b * d)
+	 */
+	for (i = 0; i < num_mic_locations - 3; i++) {
+		a = cd->mic_locations[i + 1].x - cd->mic_locations[i].x;
+		b = cd->mic_locations[i + 1].y - cd->mic_locations[i].y;
+		c = cd->mic_locations[i + 1].z - cd->mic_locations[i].z;
+		d = cd->mic_locations[i + 2].x - cd->mic_locations[i].x;
+		e = cd->mic_locations[i + 2].y - cd->mic_locations[i].y;
+		f = cd->mic_locations[i + 2].z - cd->mic_locations[i].z;
+		px = (int32_t)b * f - (int32_t)c * e;
+		py = (int32_t)c * d - (int32_t)a * f;
+		pz = (int32_t)a * e - (int32_t)b * d;
+		if (px || py || pz)
```

Comment:
I'll add here tolerance for some round error with ABS(px) > 1, so a vector component -1, 0, 1 will be treated as zero.